internal/controller/numaflowcontrollerrollout_controller.go

```diff
@@ -364,6 +374,90 @@ func (r *NumaflowControllerRolloutReconciler) getResourceOperations() (kubeUtil.
 	return ops, cleanup, nil
 }
 
+func getDeploymentCondition(status appsv1.DeploymentStatus, condType appsv1.DeploymentConditionType) *appsv1.DeploymentCondition {
+	for _, cond := range status.Conditions {
+		if cond.Type == condType {
+			return &cond
+		}
+	}
+	return nil
+}
+
+func (r *NumaflowControllerRolloutReconciler) processNumaflowControllerStatus(ctx context.Context, controllerRollout *apiv1.NumaflowControllerRollout) error {
+	numaLogger := logger.FromContext(ctx)
+
+	// Try to get the Numaflow Controller Deployment
+	deplObj := kubernetes.GenericObject{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "Deployment",
+			APIVersion: "apps/v1",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "numaflow-controller",
+			Namespace: controllerRollout.Namespace,
+		},
+	}
+	deployment, err := kubernetes.GetCR(ctx, r.restConfig, &deplObj, "deployments")
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			// This could just mean that the deployment has not yet been created.
+			// Log and return to wait for next reconcile call once the deployment is created triggering the reconciliation.
+			numaLogger.Warnf("Numaflow Controller Deployment not found. It may still have to be created.")
+			return nil
+		}
+
+		return fmt.Errorf("error getting the Numaflow Controller Deployment for NumaflowControllerRollout %s/%s: %v", controllerRollout.Namespace, controllerRollout.Name, err)
+	}
+
+	// Parse the Deployment spec of the existing Numaflow Controller Deployment
+	var deploymentSpec appsv1.DeploymentSpec
+	if len(deployment.Spec.Raw) > 0 {
+		err = json.Unmarshal(deployment.Spec.Raw, &deploymentSpec)
+		if err != nil {
+			return fmt.Errorf("unable to parse spec for existing Numaflow Controller Deployment %s/%s: %v", deployment.Namespace, deployment.Name, err)
+		}
+	}
+
+	// Parse the Deployment status of the existing Numaflow Controller Deployment
+	var deploymentStatus appsv1.DeploymentStatus
+	if len(deployment.Status.Raw) > 0 {
+		err = json.Unmarshal(deployment.Status.Raw, &deploymentStatus)
+		if err != nil {
+			return fmt.Errorf("unable to parse status for existing Numaflow Controller Deployment %s/%s: %v", deployment.Namespace, deployment.Name, err)
+		}
+	}
+
+	// Health Check borrowed from argoproj/gitops-engine/pkg/health/health_deployment.go https://github.com/argoproj/gitops-engine/blob/master/pkg/health/health_deployment.go#L27
+	if deployment.Generation <= deploymentStatus.ObservedGeneration {
```

Comment:
This `<=` comparison is probably safe to do. However, in what scenario would generation < observed generation? Due to potential k8s bugs?

Reply:
[This](https://github.com/numaproj/numaplane/pull/92#discussion_r1667084952) and [this](https://github.com/numaproj/numaplane/pull/92#discussion_r1667089759) explain why we decided to use `<=`.